Source/CarthageKit/Git.swift

```diff
@@ -13,8 +13,21 @@ import ReactiveTask
 
 /// Represents a URL for a Git remote.
 public struct GitURL: Equatable {
+
+	private let _URLString: String
+
 	/// The string representation of the URL.
-	public let URLString: String
+	public var URLString: String {
+		let gitAuth = parseGitAuthFromEnvironment()
+
+		// Do not override auth if already provided in URL
+		if let url = NSURL(string: _URLString) where url.user == nil || url.password == nil,
+			let host = url.host,
+			let auth = gitAuth[host] {
+				return _URLString.stringByReplacingOccurrencesOfString("\(host)", withString: "\(auth)@\(host)")
+		}
```

Comment:
Sorry I wasn't clearer before, but I think the appropriate place to add this logic is inside `GitHubRepository.HTTPSURL`. Then all the logic about GitHub authentication tokens can stay inside `GitHub.swift`.